contracts/cw-ics721-bridge/src/msg.rs

```diff
@@ -0,0 +1,189 @@
+    Mint {
+        /// The class_id to mint for. This must have previously been
+        /// created with `SaveClass`.
+        class_id: String,
+        /// Unique identifiers for the tokens.
+        token_ids: Vec<String>,
+        /// Urls pointing to metadata about the NFTs to mint. For
+        /// example, this may point to ERC721 metadata on IPFS. Must
+        /// be the same length as token_ids. token_uris[i] is the
+        /// metadata for token_ids[i].
+        token_uris: Vec<String>,
+        /// The address that ought to receive the NFTs. This is a
+        /// local address, not a bech32 public key.
+        receiver: String,
+    },
+    /// Much like mint, but will instantiate a new cw721 contract iff
+    /// the classID does not have one yet.
+    DoInstantiateAndMint {
+        /// The ics721 class ID to mint for.
+        class_id: String,
+        /// The URI for this class ID.
+        class_uri: Option<String>,
+        /// Unique identifiers for the tokens being transfered.
+        token_ids: Vec<String>,
+        /// A list of urls pointing to metadata about the NFTs. For
+        /// example, this may point to ERC721 metadata on ipfs.
+        ///
+        /// Must be the same length as token_ids.
+        token_uris: Vec<String>,
+        /// The address that ought to receive the NFT. This is a local
+        /// address, not a bech32 public key.
+        receiver: String,
+    },
+    /// Transfers a number of NFTs identified by CLASS_ID and
+    /// TOKEN_IDS to RECEIVER.
+    BatchTransfer {
+        /// The ics721 class ID of the tokens to be transfered.
+        class_id: String,
+        /// The address that should receive the tokens.
+        receiver: String,
+        /// The tokens (of CLASS_ID) that should be sent.
+        token_ids: Vec<String>,
+    },
+    /// Handles the falliable part of receiving an IBC
+    /// packet. Transforms TRANSFERS into a `BatchTransfer` message
+    /// and NEW_TOKENS into a `DoInstantiateAndMint`, then dispatches
+    /// those methods.
+    HandlePacketReceive {
+        /// The address receiving the NFTs.
+        receiver: String,
+        /// The URI for the collection being transfered.
+        class_uri: Option<String>,
+        /// Information about transfer actions.
+        transfers: Option<TransferInfo>,
+        /// Information about mint actions.
+        new_tokens: Option<NewTokenInfo>,
+    },
+}
+
+#[cw_serde]
+pub struct TransferInfo {
+    /// The class ID the tokens belong to.
+    pub class_id: String,
+    /// The tokens to be transfered.
+    pub token_ids: Vec<String>,
+}
+
+#[cw_serde]
+pub struct NewTokenInfo {
+    /// The class ID to mint tokens for.
+    pub class_id: String,
+    /// The token IDs of the tokens to be minted.
+    pub token_ids: Vec<String>,
+    /// The URIs of the tokens to be minted. Matched with token_ids by
+    /// index.
+    pub token_uris: Vec<String>,
+}
+
+#[cw_serde]
+pub struct IbcAwayMsg {
```

Comment:
Is this message part of the spec? Why not call it `IbcOutgoingMsg`?

Reply:
it's not part of the spec - just what the contract uses. have renamed it :)